neo4j/internal/pool/pool.go

```diff
@@ -232,16 +258,21 @@ func (p *Pool) Borrow(ctx context.Context, serverNames []string, wait bool, bolt
 	}
 
 	// Wait for a matching connection to be returned from another thread.
-	p.queueMut.Lock()
+	if !p.queueMut.TryLock(ctx) {
+		return nil, racing.LockTimeoutError("could not acquire lock in time when trying to get an idle connection")
+	}
 	// Ok, now that we own the queue we can add the item there but between getting the lock
 	// and above check for an existing connection another thread might have returned a connection
 	// so check again to avoid potentially starving this thread.
-	conn = p.tryAnyIdle(ctx, serverNames, idlenessThreshold)
+	conn, err = p.tryAnyIdle(ctx, serverNames, idlenessThreshold)
+	if err != nil {
+		return nil, err
```

Comment:
There is no release between the lock acquisition and this return => deadlock potential?

Reply:
oopsie, good catch, thanks :)